### PR TITLE
Initial serialization

### PR DESCRIPTION
**Pull request by BoskiJanusz**
This task contains:
New class FileManager for files
New class DataFileManager for data in files
New folder Database for storing data
Protobuff files
Modified tests

### CODE REVIEW

File: src/app/clinic/clinic_facade.cpp [hunk 1]
```diff
@@ -111,3 +111,19 @@ void Clinic::removeWarehouse()
...
+                        }) != doctors_.end();
+}
+
+bool Clinic::receptionistExists(const std::string& new_receptionist_pesel)
+{
+    return std::find_if(receptionists_.begin(), receptionists_.end(),
+                        [&new_receptionist_pesel](const std::shared_ptr<Receptionist>& exist_receptionists) {
+                            return new_receptionist_pesel == exist_receptionists->getPesel();
+                        }) != receptionists_.end();
+}
```
Comment (fbtom) [inline src/app/clinic/clinic_facade.cpp:129]: Could be templated in future

File: src/app/file_manager/file_manager.hpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+#include <string>
+
+class FileManager
+{
+  public:
+    enum class FileMode
+    {
+        Input,
+        OutputAppend,
+        OutputTruncate
```
Comment (fbtom) [inline src/app/file_manager/file_manager.hpp:12]: explicitly assign values

File: src/app/file_manager/paths.hpp [hunk 3]
```diff
@@ -0,0 +1,8 @@
+#pragma once
+
+namespace PATH
+{
+constexpr char DEFAULT[] = "database/";
+constexpr char WORKERS[] = "database/workers/";
+constexpr char TEST_WORKERS[] = "database/test_workers/";
```
Comment (fbtom) [inline src/app/file_manager/paths.hpp:7]: i suggest constexpr char* instead arrays

File: src/app/person/gender.cpp [hunk 4]
```diff
@@ -2,12 +2,32 @@
...
+    {
+        return Gender::Male;
+    }
+    else if (gender == "female")
+    {
+        return Gender::Female;
+    }
+    else
+        return Gender::Unknown;
+}
```
Comment (fbtom) [inline src/app/person/gender.cpp:33]: eof

File: src/app/person/gender.hpp [hunk 5]
```diff
@@ -5,7 +5,9 @@
 enum class Gender
 {
     Male,
-    Female
+    Female,
+    Unknown
```
Comment (fbtom) [inline src/app/person/gender.hpp:9]: correct to explicit values, 0/1/255? if derived from std::uint8_t

File: src/tests/ut/data_file_manager/data_file_manager_test.cpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                                               Clinic::getReceptionists().at(1));
+    test_file_content = DataFileManager::printReceptionists(PATH::TEST_WORKERS, "test_receptionists.proto");
+    expected = "Receptionist: Andrzej Szczepanski 59122991491 male\n";
+
+    DataFileManager::removeReceptionistFromFile(PATH::TEST_WORKERS, "test_receptionists.proto",
+                                                Clinic::getReceptionists().at(1));
+    test_file_content = DataFileManager::printReceptionists(PATH::TEST_WORKERS, "test_receptionists.proto");
+    expected = "";
+    EXPECT_EQ(test_file_content, expected);
+}
```
Comment (fbtom) [inline src/tests/ut/data_file_manager/data_file_manager_test.cpp:64]: eof

File: src/tests/ut/data_file_manager/data_file_manager_test.hpp [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+        {
+            Clinic::removeDoctor(doctor);
+        }
+
+        for (const auto& receptionist : Clinic::getReceptionists())
+        {
+            Clinic::removeReceptionist(receptionist);
+        }
+    }
+};
```
Comment (fbtom) [inline src/tests/ut/data_file_manager/data_file_manager_test.hpp:20]: can include test utils and call universal clinic cleaner

File: src/tests/ut/doctor/doctor_test.cpp [hunk 8]
```diff
@@ -14,9 +14,9 @@ class DoctorRemoveVisitFunction : public DoctorTestFixture
 TEST_F(DoctorTestFixture, CreatingDoctorViaCreateShouldAppendDoctorsToClinic)
 {
     Doctor::createDoctor("Jan", "Kowalski", "80062378532", Gender::Male);
-    Doctor::createDoctor("Janusz", "Schmit", "80062378532", Gender::Male);
-    Doctor::createDoctor("Gawel", "Pawel", "80062378532", Gender::Male);
-    Doctor::createDoctor("Jan", "Poniatowski", "80062378532", Gender::Male);
+    Doctor::createDoctor("Janusz", "Schmit", "70080149412", Gender::Male);
+    Doctor::createDoctor("Gawel", "Pawel", "49042481257", Gender::Male);
```
Comment (fbtom) [inline src/tests/ut/doctor/doctor_test.cpp:18]: isnt Gawel too old to work? ( ͡~ ͜ʖ ͡°)

File: src/app/person/gender.hpp [hunk 9]
```diff
@@ -5,7 +5,9 @@
...
 };
 
 std::string toString(const Gender gender);
+Gender toEnum(const std::string&);
```
Comment (fbtom) [inline src/app/person/gender.hpp:13]: are tests updated? i did not spot them

File: src/app/person/gender.cpp [hunk 10]
```diff
@@ -2,12 +2,32 @@
...
-        return "male";
+        gender_text = "male";
+        break;
 
     case Gender::Female:
-        return "female";
+        gender_text = "female";
+        break;
+    default:
+        break;
```
Comment (fbtom) [inline src/app/person/gender.cpp:16]: if we use it like that maybe its good to stay consistent with explicit assignment and unknown would be assigned

File: src/app/file_manager/file_manager.cpp [hunk 11]
```diff
@@ -0,0 +1,50 @@
+#include "file_manager.hpp"
+#include <iostream>
+#include <stdexcept>
+
+FileManager::FileManager(const std::filesystem::path& path, const std::string& file_name, FileMode mode)
```
Comment (fbtom) [inline src/app/file_manager/file_manager.cpp:5]: const filemode

File: src/app/data_file_manager/data_file_manager.hpp [hunk 12]
```diff
@@ -0,0 +1,33 @@
+#pragma once
+#include "../proto/clinic.pb.h"
```
Comment (fbtom) [inline src/app/data_file_manager/data_file_manager.hpp:2]: it is not possible to get rid of ../proto/?

File: src/app/data_file_manager/data_file_manager.cpp [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        }
+        for (int i = 0; i < receptionist_data.receptionists_size(); i++)
+        {
+            if (receptionist_data.receptionists(i).pesel() == receptionist->getPesel())
+            {
+                receptionist_data.mutable_receptionists()->DeleteSubrange(i, 1);
+                break;
+            }
+        }
+    }
```
Comment (fbtom) [inline src/app/data_file_manager/data_file_manager.cpp:95]: clever with the braces {}, i like this idea, wonder if it's considered in cpp guidlines at all :D

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

File: src/app/file_manager/paths.hpp (hunk 3 above)
Comment (BoskiJanusz) [reply src/app/file_manager/paths.hpp:7]: done

File: src/app/file_manager/file_manager.hpp (hunk 2 above)
Comment (BoskiJanusz) [reply src/app/file_manager/file_manager.hpp:12]: done

File: src/app/person/gender.hpp (hunk 9 above)
Comment (BoskiJanusz) [reply src/app/person/gender.hpp:13]: done

File: src/app/person/gender.hpp (hunk 5 above)
Comment (BoskiJanusz) [reply src/app/person/gender.hpp:9]: done

File: src/tests/ut/data_file_manager/data_file_manager_test.hpp (hunk 7 above)
Comment (BoskiJanusz) [reply src/tests/ut/data_file_manager/data_file_manager_test.hpp:20]: done

File: src/app/proto/clinic.proto [hunk 14]
```diff
@@ -0,0 +1,27 @@
+syntax = "proto3";
+
+package clinic_data;
+
+message Workers {
```
Comment (mBialczak) [inline src/app/proto/clinic.proto:5]: Minor detail: name Employees might be politically more correct.

File: src/app/person/gender.cpp (hunk 10 above)
Comment (BoskiJanusz) [reply src/app/person/gender.cpp:16]: done ( mistake )

File: src/app/file_manager/file_manager.cpp (hunk 11 above)
Comment (BoskiJanusz) [reply src/app/file_manager/file_manager.cpp:5]: done

File: src/app/data_file_manager/data_file_manager.cpp (hunk 13 above)
Comment (BoskiJanusz) [reply src/app/data_file_manager/data_file_manager.cpp:95]: I did same as here:
https://protobuf.dev/getting-started/cpptutorial/

Its safer for streams. We can consider blocking streams for safety in the future.

File: src/app/file_manager/file_manager.cpp [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+        file_.open(file_path_, std::ios::out | std::ios::app);
+        break;
+
+    case FileMode::OutputTruncate:
+        file_.open(file_path_, std::ios::out | std::ios::trunc);
+        break;
+
+    default:
+        throw std::invalid_argument{"Invalid file mode"};
+    }
```
Comment (mBialczak) [inline src/app/file_manager/file_manager.cpp:30]: seems like a perfect candidate for small helper function extraction. Switch statements have tendency to obscure code.

File: src/app/file_manager/file_manager.hpp (hunk 2 above)
Comment (mBialczak) [reply src/app/file_manager/file_manager.hpp:12]: It's a minor detail, but to me more appealing names might be something like: Read, WriteAppend, WriteTruncate or just Read, Append, Truncate

File: src/app/file_manager/file_manager.cpp [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    if (!file_.is_open())
+    {
+        throw std::runtime_error{"Cannot open file"};
+    }
+}
+
+FileManager::~FileManager()
+{
+    file_.close();
```
Comment (mBialczak) [inline src/app/file_manager/file_manager.cpp:39]: reduntant here.
fstream uses RAII and closes stream on destruction.

This makes whole constructor redundant or "default" candidate

**Review by mBialczak — COMMENTED** [review]
A hell of a good work.
I have checked only part of commits. 
But congrats.

**Comment by ziobron** [comment]
Thanks @mBialczak for Code Review.
🏅 1 XP granted

File: src/app/proto/clinic.proto (hunk 14 above)
Comment (BoskiJanusz) [reply src/app/proto/clinic.proto:5]: I will change it in next serialization part but noted :)

File: src/app/file_manager/file_manager.cpp (hunk 15 above)
Comment (BoskiJanusz) [reply src/app/file_manager/file_manager.cpp:30]: Yes indeed - done

File: src/app/data_file_manager/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,5 @@
+add_library(data_file_manager-lib data_file_manager.cpp)
+
+target_link_libraries(data_file_manager-lib PUBLIC clinic-lib file_manager-lib doctor-lib patient-lib visit-lib receptionist-lib warehouse-lib proto-lib)
+target_compile_options(data_file_manager-lib PRIVATE -Werror -Wall -Wextra -pedantic -pthread -lprotobuf -lpthread)
```
[the remark this answers is not in the document]
Comment (BoskiJanusz) [reply src/app/data_file_manager/CMakeLists.txt:4]: lpthread left

File: src/app/file_manager/file_manager.cpp (hunk 16 above)
Comment (BoskiJanusz) [reply src/app/file_manager/file_manager.cpp:39]: Done

**Review by fbtom — APPROVED** [review]

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

**Comment by ziobron** [comment]
Your PR was merged!
🏅 2 XP granted. Thanks @BoskiJanusz for making course materials better!

File: Dockerfile [hunk 18]
```diff
@@ -18,6 +20,8 @@ WORKDIR /application
 COPY /dependencies.sh /application/
 RUN /application/dependencies.sh
 
+COPY --from=BUILD /application/database /application/database
+
 COPY --from=BUILD /application/bin/${PROJECT_NAME} /application/bin/${PROJECT_NAME}
 COPY --from=BUILD /application/bin/${PROJECT_NAME}-tests /application/bin/${PROJECT_NAME}-tests
```
Comment (fbtom) [inline Dockerfile:27]: Suggestion (needs some adjustment)

# Use an official Ubuntu Mantic release as the base image for building
FROM ubuntu:mantic as BUILD

# Set environment variables
ENV PROJECT_NAME=desktop-business-app

# Set the working directory inside the container
WORKDIR /application

# Copy dependency installation script and install dependencies
# COPY dependencies.sh /application/

# Copy the rest of the application code to the working directory
COPY . /application
RUN chmod +x /application/dependencies.sh && /application/dependencies.sh

# Build the project using CMake
RUN cmake -S . -B build && cmake --build build

# Run tests
RUN /application/bin/${PROJECT_NAME}-tests

# Use a separate stage for the final image
FROM ubuntu:mantic as FINAL

# Set environment variables
ENV PROJECT_NAME=desktop-business-app

# Set the working directory inside the container
WORKDIR /application
COPY . /application
# Copy dependency installation script and install dependencies
# COPY dependencies.sh /application/
RUN chmod +x /application/dependencies.sh && /application/dependencies.sh

# Copy only the necessary files from the build stage
COPY --from=BUILD /application/build/${PROJECT_NAME} /application/bin/${PROJECT_NAME}
COPY --from=BUILD /application/build/${PROJECT_NAME}-tests /application/bin/${PROJECT_NAME}-tests

# Set the entry point for the container
ENTRYPOINT ["/application/bin/desktop-business-app"]

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

**Pull request merged**